### ISSUE TITLE
[neural.lua] Stack trace while learning ANN

### ISSUE THREAD

**Issue by HansVanEijsden** [rspamd/rspamd#2793]
### Classification (Please choose *one* option):

* [X] Crash/Hang/Data loss
* [ ] WebUI/Usability
* [ ] Unintended behaviour (bug)
* [ ] Enhancement

### Reproducibility (Please choose *one* option):

* [X] Always
* [ ] Sometimes
* [ ] Rarely
* [ ] Unable
* [ ] I didn’t try
* [ ] Not applicable

### Rspamd version:
1.9.0 (1.9.0-3~stretch)

### Operation system (the exact version), CPU (including architecture and processor family):
Debian GNU/Linux 9, Linux vps 4.19.0-0.bpo.2-amd64 #1 SMP Debian 4.19.16-1~bpo9+1 (2019-02-07) x86_64 GNU/Linux

### Description (Please provide a descriptive summary of the issue):
As soon as Rspamd wants to learn ANN neural networks, the process crashes and hangs.

### Relevant logs (see details [here](https://rspamd.com/doc/faq.html#how-to-debug-some-module-in-rspamd)):
[rspamd-log.txt](https://github.com/rspamd/rspamd/files/2966155/rspamd-log.txt)
I also just enabled debugging for the Neural module, there's not much more extra debug information. I'll add more results if they will be there.

### Expected results:
Learning both short and medium ANNs, without errors and without hangs or crashes

### Actual results:
Neural learning processes are hanging/crashing while learning short and medium ANNs

### Configuration (e.g. `rspamadm configdump module`):
```
# rspamadm configdump neural
*** Section neural ***
rules {
    LONG {
        train {
            learning_rate = 0.010000;
            max_usages = 2000;
            spam_score = 8;
            ham_score = -2;
            max_trains = 1000;
            max_iterations = 25;
        }
        symbol_spam = "NEURAL_SPAM_LONG";
        symbol_ham = "NEURAL_HAM_LONG";
        ann_expire = 8640000;
    }
    MEDIUM {
        train {
            learning_rate = 0.010000;
            max_usages = 20;
            spam_score = 8;
            ham_score = -2;
            max_trains = 100;
            max_iterations = 25;
        }
        symbol_spam = "NEURAL_SPAM_MEDIUM";
        symbol_ham = "NEURAL_HAM_MEDIUM";
        ann_expire = 864000;
    }
    SHORT {
        train {
            learning_rate = 0.010000;
            max_usages = 5;
            spam_score = 8;
            ham_score = -2;
            max_trains = 50;
            max_iterations = 25;
        }
        symbol_spam = "NEURAL_SPAM_SHORT";
        symbol_ham = "NEURAL_HAM_SHORT";
        ann_expire = 259200;
    }
}
train {
    ham_score = -2;
    max_usages = 20;
    spam_score = 8;
    learning_rate = 0.010000;
    max_iterations = 25;
    max_train = 1000;
}
enabled = "yes";
timeout = 20;
use_settings = false;

*** End of section neural ***
```

### Additional information:
Training of the first and second short ANN went well. After that, problems began to appear.
Restarting Rspamd didn't solve it.

I also saw this, don't now if it's related:
```
2019-03-14 05:52:09 #11668(rspamd_proxy) <de160e>; proxy; insert_metric_result: detected NaN score for symbol NEURAL_HAM_SHORT, replace it with zero
2019-03-14 07:09:11 #11668(rspamd_proxy) <34d171>; proxy; insert_metric_result: detected NaN score for symbol NEURAL_HAM_SHORT, replace it with zero
2019-03-14 07:21:36 #11668(rspamd_proxy) <5213b5>; proxy; insert_metric_result: detected NaN score for symbol NEURAL_HAM_SHORT, replace it with zero
```

**Comment by csehvendel** [comment]
Same problem in same envoriment....

**Comment by HansVanEijsden** [comment]
@csehvendel I had the problems on 47/47 servers, on all the servers I had to erase all the neural learning vectors from the Redis database and I had to start all over again. Until now it works (knock on wood). More information about how to remove all the keys from Redis (not documented yet): 
https://lists.rspamd.com/pipermail/users/2019-March/000430.html

**Comment by csehvendel** [comment]
@HansVanEijsden Thanks, it doesn't work for me ...

**Comment by HansVanEijsden** [comment]
@csehvendel what happens if you give the rules other names (so it really starts over 100% clean)? For example, `LONG {` -> `LONGER {` and `SHORT {` -> `SHORTER {`?

**Comment by csehvendel** [comment]
@HansVanEijsden mutch better! :)
rules {
    "MOUNTHLY" {
        ......
        ann_expire = 30d;
    }
    "WEEKLY" {
        .....
        ann_expire = 7d;
    }
    "DAILY" {
        .....
        ann_expire = 1d;
    }
}

**Comment by csehvendel** [comment]
The miracle is over.....

**Comment by Kuzya80** [comment]
Similar problem...

Mar 20 12:26:48 rspamd[14826]: <>; ; rspamd_lua_execute_lua_subprocess: call to subprocess failed: /usr/share/rspamd/lualib/nn/Container.lua:67: #012In 1 module of nn.Sequential:#012/usr/share/rspamd/lualib/nn/NaN.lua:50: NaN found in gradOutput of module :#012nn.NaN(1) @ nn.Identity#012stack traceback:#012#011[C]: in function 'error'#012#011/usr/share/rspamd/lualib/nn/NaN.lua:50: in function </usr/share/rspamd/lualib/nn/NaN.lua:46>#012#011[C]: in function 'xpcall'#012#011/usr/share/rspamd/lualib/nn/Container.lua:63: in function 'rethrowErrors'#012#011/usr/share/rspamd/lualib/nn/Sequential.lua:58: in function 'updateGradInput'#012#011/usr/share/rspamd/lualib/nn/StochasticGradient.lua:37: in function 'train'#012#011/usr/share/rspamd/plugins/neural.lua:676: in function </usr/share/rspamd/plugins/neural.lua:661>#012#011[C]: in function 'spawn_process'#012#011/usr/share/rspamd/plugins/neural.lua:685: in function 'callback'#012#011/usr/share/rspamd/lualib/lua_redis.lua:921: in function </usr/share/rspamd/lualib/lua_redis.lua:915>#012#012WARNING: If you see a stack trace below, it doesn't point to the place where this error occurred. Please use only the one above.; trace: [1]:{[C]:-1 - error [C]}; [2]:{/usr/share/rspamd/lualib/nn/Container.lua:67 - rethrowErrors [Lua]}; [3]:{/usr/share/rspamd/lualib/nn/Sequential.lua:58 - updateGradInput [Lua]}; [4]:{/usr/share/rspamd/lualib/nn/StochasticGradient.lua:37 - train [Lua]}; [5]:{/usr/share/rspamd/plugins/neural.lua:676 - <unknown> [Lua]}; [6]:{[C]:-1 - spawn_process [C]}; [7]:{/usr/share/rspamd/plugins/neural.lua:685 - callback [Lua]}; [8]:{/usr/share/rspamd/lualib/lua_redis.lua:921 - <unknown> [Lua]};

**Comment by csehvendel** [comment]
Same situation with single neural network.

**Comment by csehvendel** [comment]
Same problem with fresh install!
Any ideas?

**Comment by HansVanEijsden** [comment]
The issue is back, also here. I don't have any ideas unfortunately, I just can't get it working: after several days things still go nuts.

**Comment by skroczek** [comment]
My installation is also affected by this behaviour (Debian stretch; rspamd 1.9.0-3~stretch). The only thing I was able to investigate, is, that the dataset seems to be somehow empty when the error occours:

In `/usr/share/rspamd/plugins/neural.lua` line 653
this `s` variable with content
```
{[1] = 0, [2] = 0, [3] = 0, [4] = 0, [5] = 0, [6] = 0, [7] = 0, [8] = 0, [9] = 0, [10] = 0, [11] = 0, [12] = 0, [13] = 0, [14] = 0, [15] = 0, [16] = 0, [17] = 0, [18] = 0, [19] = 0, [20] = 0, [21] = 0, [22] = 0, [23] = 0, [24] = 0, [25] = 0, [26] = 0, [27] = 0, [28] = 0, [29] = 0, [30] = 0, [31] = 0, [32] = 0, [33] = 0, [34] = 0, [35] = 0, [36] = 0, [37] = 0, [38] = 0, [39] = 0, [40] = 0, [41] = 0, [42] = 0, [43] = 0, [44] = 0, [45] = 0, [46] = 0, [47] = 0, [48] = 0, [49] = 0, [50] = 0, [51] = 0, [52] = 0, [53] = 0, [54] = 0, [55] = 0, [56] = 0, [57] = 0, [58] = 0, [59] = 0, [60] = 0, [61] = 0, [62] = 0, [63] = 0, [64] = 0, [65] = 0, [66] = 0, [67] = 0, [68] = 0, [69] = 0, [70] = 0, [71] = 0, [72] = 0, [73] = 0, [74] = 0, [75] = 0, [76] = 0, [77] = 0, [78] = 0, [79] = 0, [80] = 0, [81] = 0, [82] = 0.761594, [83] = 0, [84] = 0, [85] = 0, [86] = 0, [87] = 0, [88] = 0, [89] = 0, [90] = 0, [91] = 0, [92] = 0, [93] = 0, [94] = 0, [95] = 0, [96] = 0, [97] = 0, [98] = 0, [99] = 0, [100] = 0, [101] = 0, [102] = 0, [103] = 0, [104] = 0, [105] = 0, [106] = 0, [107] = 0, [108] = 0, [109] = 0, [110] = 0, [111] = 0, [112] = 0, [113] = 0, [114] = 0, [115] = 0, [116] = 0, [117] = -0.099668, [118] = 0, [119] = 0, [120] = 0, [121] = 0, [122] = 0, [123] = 0, [124] = 0, [125] = 0, [126] = 0, [127] = 0, [128] = 0, [129] = 0, [130] = 0, [131] = 0, [132] = 0, [133] = 0, [134] = 0, [135] = 0, [136] = 0, [137] = 0, [138] = 0, [139] = 0, [140] = 0, [141] = 0, [142] = 0, [143] = 0, [144] = 0, [145] = 0, [146] = 0, [147] = 0, [148] = 0, [149] = 0, [150] = 0, [151] = 0, [152] = 0, [153] = 0, [154] = 0, [155] = 0, [156] = 0, [157] = 0, [158] = 0, [159] = 0, [160] = 0, [161] = 0, [162] = 0, [163] = -0.998878, [164] = 0, [165] = 0, [166] = 0, [167] = 0, [168] = 0, [169] = 0, [170] = 0, [171] = 0, [172] = 0, [173] = 0, [174] = 0, [175] = 0, [176] = 0, [177] = 0, [178] = 0, [179] = 0, [180] = 0, [181] = 0, [182] = 0, [183] = 0, [184] = 0, [185] = 0, [186] = 0, [187] = 0, [188] = 0, [189] = 0, [190] = 0, [191] = 0, [192] = 0, [193] = 0, [194] = 0, [195] = 0, [196] = 0, [197] = 0, [198] = 0, [199] = 0, [200] = 0, [201] = 0, [202] = 0, [203] = 0, [204] = 0, [205] = 0, [206] = 0, [207] = 0, [208] = 0, [209] = 0, [210] = 0, [211] = 0, [212] = 0, [213] = 0, [214] = 0, [215] = 0, [216] = 0, [217] = 0, [218] = 0, [219] = 0, [220] = 0, [221] = 0, [222] = 0, [223] = 0, [224] = 0, [225] = 0, [226] = 0, [227] = 0, [228] = 0, [229] = 0, [230] = 0, [231] = 0.379949, [232] = 0, [233] = 0, [234] = 0, [235] = 0, [236] = 0, [237] = 0, [238] = 0, [239] = 0, [240] = 0, [241] = 0, [242] = 0, [243] = 0, [244] = 0, [245] = 0, [246] = 0, [247] = 0, [248] = 0, [249] = 0, [250] = 0, [251] = 0, [252] = 0, [253] = 0, [254] = 0, [255] = 0, [256] = 0, [257] = 0, [258] = 0, [259] = 0, [260] = 0, [261] = 0, [262] = 0, [263] = 0, [264] = 0, [265] = 0, [266] = 0, [267] = 0, [268] = 0, [269] = 0, [270] = 0, [271] = 0, [272] = 0, [273] = 0, [274] = 0, [275] = 0, [276] = 0, [277] = 0, [278] = 0, [279] = 0, [280] = 0, [281] = 0, [282] = 0, [283] = 0, [284] = 0, [285] = 0, [286] = 0, [287] = 0, [288] = 0, [289] = 0, [290] = 0, [291] = 0, [292] = 0, [293] = 0, [294] = 0, [295] = 0, [296] = 0, [297] = 0, [298] = 0, [299] = 0, [300] = 0, [301] = 0, [302] = 0, [303] = 0, [304] = 0, [305] = 0, [306] = 0, [307] = 0, [308] = 0, [309] = 0, [310] = 0, [311] = 0, [312] = 0, [313] = 0, [314] = 0, [315] = 0, [316] = 0, [317] = 0, [318] = 0, [319] = 0, [320] = 0, [321] = 0, [322] = 0, [323] = 0, [324] = 0, [325] = 0, [326] = 0, [327] = 0, [328] = 0, [329] = 0, [330] = 0, [331] = 0, [332] = 0, [333] = 0, [334] = 0, [335] = 0, [336] = 0, [337] = 0, [338] = 0, [339] = 0, [340] = 0, [341] = 0, [342] = 0, [343] = 0, [344] = 0, [345] = 0, [346] = 0, [347] = 0, [348] = 0, [349] = 0, [350] = 0, [351] = 0, [352] = 0, [353] = 0, [354] = 0, [355] = 0, [356] = 0, [357] = 0, [358] = 0, [359] = 0, [360] = 0, [361] = 0, [362] = 0, [363] = 0, [364] = 0, [365] = 0, [366] = 0, [367] = 0, [368] = 0, [369] = 0, [370] = 0, [371] = 0, [372] = 0, [373] = 0, [374] = 0, [375] = 0, [376] = 0, [377] = 0, [378] = 0, [379] = 0, [380] = 0, [381] = 0, [382] = 0, [383] = 0, [384] = 0, [385] = 0, [386] = 0, [387] = 0, [388] = 0, [389] = 0, [390] = 0, [391] = 0, [392] = 0, [393] = 0, [394] = 0, [395] = 0, [396] = 0, [397] = 0, [398] = 0, [399] = 0, [400] = 0, [401] = 0, [402] = 0, [403] = 0, [404] = 0, [405] = 0, [406] = 0, [407] = 0, [408] = 0, [409] = 0, [410] = 0, [411] = 0, [412] = 0, [413] = 0, [414] = 0, [415] = 0, [416] = 0, [417] = 0, [418] = 0, [419] = 0, [420] = 0, [421] = 0, [422] = 0, [423] = 0, [424] = 0, [425] = 0, [426] = 0, [427] = 0, [428] = 0, [429] = 0, [430] = 0, [431] = 0, [432] = 0, [433] = 0, [434] = 0, [435] = 0, [436] = 0, [437] = 0, [438] = 0, [439] = 0, [440] = 0, [441] = 0, [442] = 0, [443] = 0, [444] = 0, [445] = 0, [446] = 0, [447] = 0, [448] = 0, [449] = 0, [450] = 0, [451] = 0, [452] = 0, [453] = 0, [454] = 0, [455] = 0, [456] = 0, [457] = 0, [458] = 0, [459] = 0, [460] = 0, [461] = 0, [462] = 0, [463] = 0, [464] = 0, [465] = 0, [466] = 0, [467] = 0, [468] = 0, [469] = 0, [470] = 0, [471] = 0, [472] = 0, [473] = 0, [474] = 0, [475] = 0, [476] = 0, [477] = 0, [478] = 0.250000, [479] = 0, [480] = 0, [481] = 0, [482] = 0, [483] = 0, [484] = 0.500000, [485] = 0, [486] = 1, [487] = 0, [488] = 1, [489] = 1, [490] = 0, [491] = 0, [492] = 0, [493] = 0, [494] = 1, [495] = 0.154297, [496] = 0.090909, [497] = 0, [498] = 0, [499] = 0.932384, [500] = 0.932384, [501] = 0, [502] = 0.085409, [503] = 0.128114}
```
becomes
```
{[1] = , [2] = }
```
after foreach. But I have no idea if this is expected or not...

**Comment by vstakhov** [comment]
I don't understand what variable do you print.

**Comment by vstakhov** [comment]
Actually, I was able to reproduce this problem and added lots of debugging there. But it all turns out that some of the gradients in linear input layer are NaN. I've tried to collect all data and trained the same model locally. Apparently it was fine... 

So I have no ideas about what to do with the torch. Another issue is that this project is now stale and its code quality is extremely low - it's more like it was written by a group of students than by professional software developers.

**Comment by HansVanEijsden** [comment]
@vstakhov too bad about the code quality, I wish I was a coder so I could help, but unfortunately I'm not that skilled. I would like to thank you for trying to reproduce the problem and adding lots of debugging.
Let's hope for a better future, because the results of the Neural module were really great here: no false positives and really better scores.

**Comment by vstakhov** [comment]
It seems I have found the issue. Some of the symbols had infinity value.

**Comment by vstakhov** [comment]
It should be fixed now.

**Comment by HansVanEijsden** [comment]
You really made my day. 🎉
I replaced `neural.lua` in `/usr/share/rspamd/plugins` with the most recent one, 781f316, cleared Redis and just to be sure I changed the neural rules names: a fresh start.
Let's see how it goes, but it feels good! 😎

**Comment by HansVanEijsden** [comment]
@vstakhov I just sent you a donation via PayPal, so you can buy yourself some good beer, wine or pizza. 👍🏼
